File: dart/lib/src/packets/src/client/pc.dart
part of '../../packets.dart';

class PcPacket extends ClientPacket {
  /// [timestamp] is the time of the package.
  /// This is identified in the package as `UNIX`
  final DateTime timestamp;

  /// [commandId] is the command ID that is being ACKed.
  /// This is identified in the package as `CMD_ID`
  final String commandId;

  /// [message] is the message of the ACK.
  /// This is identified in the package as `MSG`
  final String message;

  /// [PcPacket] is the command ACK package.
  ///
  /// This package is part of the package sent from the device to the server.
  ///
  /// This package should be send after the execution of a command, received in [Ac] package.
  PcPacket({
    required this.timestamp,
    required this.commandId,
    required this.message,
  });

  /// [fromPacket] creates a [PcPacket] from a string package in the format of `Layrz Protocol v2`.
  static PcPacket fromPacket(String raw) {
    if (!raw.startsWith('<Pc>') || !raw.endsWith('</Pc>')) {
      throw ParseException('Invalid identification package, should be <Pc>...</Pc>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    if (parts.length != 4) {
      throw MalformedException('Invalid package parts, should have 4 parts');
    }

    int? receivedCrc = int.tryParse(parts[3], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 3).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    DateTime timestamp;
    try {
      timestamp = DateTime.fromMillisecondsSinceEpoch(int.parse(parts[0]) * 1000);
    } catch (e) {
      throw MalformedException('Invalid timestamp');
    }

    return PcPacket(
      timestamp: timestamp,
      commandId: parts[1],
      message: parts[2],
    );
  }

  /// [toPacket] returns the package in the format of `Layrz Protocol v2`.
  @override
  String toPacket() {
    String payload = '${(timestamp.millisecondsSinceEpoch / 1000).round()};$commandId;$message;';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();

    return '<Pc>$payload$crc</Pc>';
  }

  @override
  String toString() => toPacket();

  @override
  PcPacket copyWith({
    DateTime? timestamp,
    String? commandId,
    String? message,
  }) {
    return PcPacket(
      timestamp: timestamp ?? this.timestamp,
      commandId: commandId ?? this.commandId,
      message: message ?? this.message,
    );
  }
}

File: go/pc.go
package layrzprotocol

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Based on the `Layrz Protocol v2` specification. PcPacket is the Packet command response
// sent from the device to the server
type PcPacket struct {
	// Is the timestamp of the response packet
	Timestamp time.Time

	// Is the command id of the response packet
	CommandId int

	// Is the message of the response packet
	Message *string
}

// FromPacket is a method that converts a raw packet to a PaPacket
// based on the `Layrz Protocol v2` specification
//
// Returns an error if the packet is invalid
func (p *PcPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Pc>") || !strings.HasSuffix(*raw, "</Pc>") {
		return errors.New("invalid package, should be <Pc>...</Pc>")
	}

	*raw = strings.TrimPrefix(*raw, "<Pc>")
	*raw = strings.TrimSuffix(*raw, "</Pc>")

	rawCrc := (*raw)[len(*raw)-4:]
	*raw = (*raw)[:len(*raw)-4]

	receivedCrc, err := strconv.ParseUint(rawCrc, 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := calculateCrc([]byte(*raw))

	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	*raw = strings.TrimSuffix(*raw, ";")

	parts := strings.Split(*raw, ";")

	if len(parts) != 3 {
		return errors.New("invalid package, should contain 3 parts")
	}

	rawTimestamp, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return errors.New("cannot convert timestamp to integer")
	}

	p.Timestamp = time.Unix(rawTimestamp, 0)
	p.Message = &parts[2]

	p.CommandId, err = strconv.Atoi(parts[1])
	if err != nil {
		return errors.New("cannot convert command id to integer")
	}

	return nil
}

// ToPacket is a method that converts a PaPacket to a raw packet
// based on the `Layrz Protocol v2` specification
func (p *PcPacket) ToPacket() *string {
	content := fmt.Sprintf("%d;%d;%s;", p.Timestamp.Unix(), p.CommandId, *p.Message)
	crc := calculateCrc([]byte(content))
	content = fmt.Sprintf("<Pc>%s%04X</Pc>", content, crc)
	return &content
}
